adapter/internal/discovery/xds/marshaller.go

```diff
@@ -380,3 +328,326 @@ func MarshalKeyManager(keyManager *types.KeyManager) *keymgt.KeyManagerConfig {
+	}
+	ApplicationMap = resourceMap
+	updateEnforcerApplications(marshalApplicationMapToList(ApplicationMap))
+}
+
+// UpdateXdsForApplicationEvent handles the Application Event corresponding to the event received
+// from message broker. In the end, XDS cache will be updated as well.
+func UpdateXdsForApplicationEvent(application *types.Application, eventType EventType) {
+	if eventType == DeleteEvent {
+		delete(ApplicationMap, application.UUID)
+		logger.LoggerXds.Infof("Application %s is deleted.", application.UUID)
+	} else {
+		applicationSub := marshalApplication(application)
+		ApplicationMap[application.UUID] = applicationSub
+		if eventType == CreateEvent {
+			logger.LoggerXds.Infof("Application %s is added.", application.UUID)
+		} else {
+			logger.LoggerXds.Infof("Application %s is updated.", application.UUID)
+		}
+	}
+	updateEnforcerApplications(marshalApplicationMapToList(ApplicationMap))
+}
+
+// UpdateXdsWithMultipleApplicationKeyMappings is used to update the application key mappings during the startup where
+// multiple key mappings are pulled at once. In the end, XDS cache will be updated as well.
+func UpdateXdsWithMultipleApplicationKeyMappings(keymappingList *types.ApplicationKeyMappingList) {
+	resourceMap := make(map[string]*subscription.ApplicationKeyMapping)
+	for _, keyMapping := range keymappingList.List {
+		applicationKeyMappingReference := GetApplicationKeyMappingReference(&keyMapping)
+		keyMappingSub := marshalKeyMapping(&keyMapping)
+		resourceMap[applicationKeyMappingReference] = keyMappingSub
+	}
+	ApplicationKeyMappingMap = resourceMap
+	updateEnforcerApplicationKeyMappings(marshalKeyMappingMapToList(ApplicationKeyMappingMap))
+}
+
+// UpdateXdsForApplicationKeyMappingEvent handles the Application Key Mapping Event corresponding to the event received
+// from message broker. In the end, XDS cache will be updated as well.
+func UpdateXdsForApplicationKeyMappingEvent(keyMapping *types.ApplicationKeyMapping, eventType EventType) {
+	applicationKeyMappingReference := GetApplicationKeyMappingReference(keyMapping)
+	if eventType == DeleteEvent {
+		delete(ApplicationKeyMappingMap, applicationKeyMappingReference)
+		logger.LoggerXds.Infof("Application Key Mapping for the applicationKeyMappingReference %s is removed.",
+			applicationKeyMappingReference)
+	} else {
+		keyMappingSub := marshalKeyMapping(keyMapping)
+		ApplicationKeyMappingMap[applicationKeyMappingReference] = keyMappingSub
+		logger.LoggerXds.Infof("Application Key Mapping for the applicationKeyMappingReference %s is added.",
+			applicationKeyMappingReference)
+	}
+	updateEnforcerApplicationKeyMappings(marshalKeyMappingMapToList(ApplicationKeyMappingMap))
+}
+
+// UpdateXdsForMultipleSubscriptions is used to update the subscriptions during the startup where
+// multiple subscriptions are pulled at once. In the end, XDS cache will be updated as well.
+func UpdateXdsForMultipleSubscriptions(subscriptionsList *types.SubscriptionList) {
```

Comment:
better to use this file only for "marshalling" related functions. basically here you should only marshall the required objects and return. another party should call the update xds methods.  updating xds shouldn't be a part of marshalling logic's responsibility. Same for all other new `UpdateXdsFor...` methods introduced in here